aggregator/src/aggregator/key_rotator.rs

```diff
@@ -0,0 +1,266 @@
+    hpke::{generate_hpke_config_and_private_key, HpkeCiphersuite, HpkeKeypair},
+    time::{Clock, TimeExt},
+};
+use janus_messages::{Duration, HpkeConfigId};
+use serde::{Deserialize, Serialize};
+
+#[derive(Debug)]
+pub struct KeyRotator<C: Clock> {
+    clock: C,
+    datastore: Arc<Datastore<C>>,
+    hpke: HpkeKeyRotatorConfig,
+}
+
+/// Handles key rotation for global HPKE keys. It is assumed that the [`KeyRotator`] is the sole
+/// entity responsible for the `global_hpke_keys` table. Do not make manual changes to the
+/// `global_hpke_keys` table when the key rotator is used.
+#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
+pub struct HpkeKeyRotatorConfig {
+    /// How long key remains in [`HpkeKeyState::Pending`] before being moved to
+    /// [`HpkeKeyState::Active`]. This should correlate to
+    /// [`AggregatorConfig::global_hpke_configs_refresh_interval`]
+    pub pending_duration: Duration,
+
+    /// The time-to-live of the key. Once this is exceeded, the key is moved to
+    /// [`HpkeKeyState::Expired`]. It is at operator discretion as to how long this should be.
+    pub age: Duration,
+
+    /// How long the key remains in [`HpkeKeyState::Expired`] before being deleted. This should
+    /// correlate to how long clients are expected to cache an HPKE key.
+    pub expired_duration: Duration,
+
+    /// List of ciphersuites to rotate. The key rotator handles keys that use each specified
+    /// ciphersuite.
+    pub ciphersuites: Vec<HpkeCiphersuite>,
+}
+
+impl<C: Clock> KeyRotator<C> {
+    pub fn new(clock: C, datastore: Arc<Datastore<C>>, hpke: HpkeKeyRotatorConfig) -> Self {
+        Self {
+            clock,
+            datastore,
+            hpke,
+        }
+    }
+
+    pub async fn run(&self) -> Result<(), Error> {
+        self.datastore
+            .run_tx("global_hpke_key_rotator", |tx| {
+                let clock = self.clock.clone();
+                let hpke_key_rotator_config = self.hpke.clone();
+                Box::pin(async move {
+                    // Take an ExclusiveLock lock on the table. This ensures that only one key
+                    // rotator replica writes to the table at a time and that each replica gets
+                    // a consistent view of the table state.
+                    //
+                    // ExclusiveLock does not conflict with the AccessShare lock taken by table
+                    // reads, i.e. other aggregator replicas can continue to refresh their key
+                    // caches without being blocked.
+                    tx.lock_global_hpke_keypairs().await?;
+
+                    let current_keypairs = tx.get_global_hpke_keypairs().await?;
+
+                    let ids: Vec<_> = current_keypairs
+                        .iter()
+                        .map(|keypair| u8::from(*keypair.hpke_keypair().config().id()))
+                        .collect();
+                    let mut available_ids = (0..=u8::MAX).filter_map(|id| {
```

Comment:
This is less important, but we could ensure that our IDs cycle through the whole u8 space if we identified which keypair was created most recently, and then replaced `(0..=u8::MAX)` with `(newest_id..=u8::MAX).chain(0..newest_id)`. That way, weird clients with very old cached keys will have a much bigger window during which we will send them urn:ietf:params:ppm:dap:error:outdatedConfig, instead of trying to decrypt it with a new key with the same ID.